Run a validator (Polkadex) (#301)
Polkadex-Substrate/Polkadex: docs/run-a-validator.md

```
sudo ntpq -p
```

### Building and Installing the `Polkadex` binary

>If you don't want to build the binary and simply prefer to download it, use the following command.
>`curl -o polkadex-node https://github.com/Polkadex-Substrate/Polkadex/releases/download/v0.4.0/polkadex-node`

You will need to build the `Polkadex` binary from the [Polkadex-Substrate/Polkadex](https://github.com/Polkadex-Substrate/Polkadex) repository on GitHub using the source code available in the v0.4.0 release.

```
cd $HOME
git clone https://github.com/Polkadex-Substrate/Polkadex.git
cd Polkadex
```

Build native code with the cargo release profile.

```
git checkout v0.4.0
rustup toolchain add nightly-2021-05-11
rustup target add wasm32-unknown-unknown --toolchain nightly-2021-05-11
rustup target add x86_64-unknown-linux-gnu --toolchain nightly-2021-05-11
cargo build --release
```

<i>This step will take a while (generally 10 - 40 minutes, depending on your hardware).</i>

### Synchronize Chain Data

Download `customSpecRaw.json` file for the Polkadex Public Testnet
```
cd $HOME
curl -o customSpecRaw.json https://github.com/Polkadex-Substrate/Polkadex/releases/download/v0.4.0/customSpecRaw.json
```

> You can begin syncing your node by running the following command:

Reviewer: ```suggestion
You can begin syncing your node by running the following command if you do not want to start in validator mode right away:
```

```
$HOME/Polkadex/target/release/polkadex-node --chain=$HOME/customSpecRaw.json --bootnodes /ip4/13.235.92.50/tcp/30333/p2p/12D3KooWBRsL9KPkMeWxTMq5aSbgUWEMgwzWpWDA6EqQ6A2KTDoR --pruning=archive
```

> if you do not want to start in validator mode right away.

Reviewer: ```suggestion

```

- Select the Controller. This is the account that will decide when to start or stop validating.

First, go to the [Staking](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fblockchain.polkadex.trade#/staking/actions) section. Click on "Account Actions", and then the "+ Stash" button.

![Bonding Preferences](./screenshots/bonding_preferences.png)

- <strong>Stash account</strong> - Select your Stash account. In this example, we will bond 100 PDEX, where the minimum bonding amount is 100. Make sure that your Stash account contains at least this much. You can, of course, stake more than this.
- <strong>Controller account</strong> - Select the Controller account created earlier. This account will also need a small amount of PDEX in order to start and stop validating.
- <strong>Value bonded</strong> - How much PDEX from the Stash account you want to bond/stake. Note that you do not need to bond all of the PDEX in that account. Also note that you can always bond more PDEX later. However, withdrawing any bonded amount requires the duration of the unbonding period. On Polkadex, the planned unbonding period is 28 days.

> - <strong>Payment destination</strong> - The account where the rewards from validating are sent. More info here. If you'd like to redirect payments to an account that is neither the controller nor the stash account, set one up. Please note that it is extremely unsafe to set an exchange deposit address as the recipient of the staking rewards.

Reviewer: "More info here." delete 
```suggestion
- <strong>Payment destination</strong> - The account where the rewards from validating are sent. If you'd like to redirect payments to an account that is neither the controller nor the stash account, set one up. Please note that it is extremely unsafe to set an exchange deposit address as the recipient of the staking rewards.
```

- <strong>CPU</strong> - Intel(R) Core(TM) i7-7700K CPU @ 4.20GHz
- <strong>Storage</strong> - A NVMe solid state drive. Should be reasonably sized to deal with blockchain growth. Starting around 80GB - 160GB will be okay for the first six months of Polkadex, but will need to be re-evaluated every six months.
- <strong>Memory</strong> - 64GB

The specs posted above are by no means the minimum specs that you could use when running a validator, however you should be aware that if you are using less you may need to toggle some extra optimizations in order to be equal to other validators that are running the standard.

### Node Prerequisites: Install Rust and Dependencies

Once you choose your cloud service provider and set-up your new server, the first thing you will do is install Rust.

If you have never installed Rust, you should do this first.

If you have already installed Rust, run the following command to make sure you are using the latest version.

```
rustup update
```

If not, this command will fetch the latest version of Rust and install it.

```
curl https://sh.rustup.rs -sSf | sh -s -- -y
```

> > If you do not have "curl" installed, run "sudo apt install curl"

Reviewer 2: Put curl in the sudo apt-get install section

To configure your shell, run the following command.

```
source $HOME/.cargo/env
```

Verify your installation.

```
rustc --version
```

Finally, run this command to install the necessary dependencies for compiling and running the Polkadex node.

```
sudo apt install make clang pkg-config libssl-dev build-essential

> ```

Reviewer 2: Move this before rustup

If you are using Ubuntu 18.04 / 20.04, NTP Client should be installed by default.

```
timedatectl
```

If NTP is installed and running, you should see `System clock synchronized: yes` (or a similar message). If you do not see it, you can install it by executing:

```
sudo apt-get install ntp
```

ntpd will be started automatically after install. You can query ntpd for status information to verify that everything is working:

```
sudo ntpq -p
```

### Building and Installing the `Polkadex` binary

You will need to build the `Polkadex` binary from the [Polkadex-Substrate/Polkadex](https://github.com/Polkadex-Substrate/Polkadex) repository on GitHub using the source code available in the v0.4.0 release.

```
git clone https://github.com/Polkadex-Substrate/Polkadex.git
cd Polkadex
```

Build native code with the cargo release profile.

```

> git checkout v0.4.0

Reviewer 2: tag not yes released - known issue

> ./scripts/init.sh

Reviewer 2: Must be updated only fetches wasm for nightly not nightly 20210511 which we are pinned to

This guide will instruct you how to set up a validator node on the Polkadex Public Testnet.

- **Github**: https://github.com/Polkadex-Substrate/Polkadex/releases/tag/v0.4.0
- **Bootnode IP address**: 13.235.92.50
- **Bootnode Peer ID**: 12D3KooWBRsL9KPkMeWxTMq5aSbgUWEMgwzWpWDA6EqQ6A2KTDoR
- **customSpecRaw.json** https://github.com/Polkadex-Substrate/Polkadex/releases/download/v0.4.0/customSpecRaw.json

## Initial Set-up

### Requirements

The most common way for a beginner to run a validator is on a cloud server running Linux. You may choose whatever VPS provider you prefer, and whichever operating system you are comfortable with. For this guide we will be using **Ubuntu 20.04**, but the instructions should be similar for other platforms.

> The transactions weights in Polkadot were benchmarked on standard hardware. It is recommended that validators run at least the standard hardware in order to ensure they are able to process all blocks in time. The following are not minimum requirements but if you decide to run with less than this, you may experience performance issues.

Reviewer 3: Transaction weights

Reviewer 2: ```suggestion
The transaction weights in Polkadot were benchmarked on standard hardware. It is recommended that validators run at least the standard hardware in order to ensure they are able to process all blocks in time. The following are not minimum requirements but if you decide to run with less than this, you may experience performance issues.
```

curl https://sh.rustup.rs -sSf | sh -s -- -y
```

> To to build you need to add some toolschains 

Reviewer 3: 'To' is used twice.

Reviewer 2: ```suggestion
Add the required toolchains with rustup
```

If you are interested in determining how much longer you have to go, your server logs (printed to STDOUT from the `Polkadex` process) will tell you the latest block your node has processed and verified. You can then compare that to the current highest block via [Telemetry](https://telemetry.polkadot.io/#list/Polkadex%20Public%20Testnet) or the [PolkadotJS Block Explorer](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fblockchain.polkadex.trade#/explorer).

## Bond PDEX

> [Create two accounts](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fblockchain.polkadex.trade#/accounts) [Here is an instruction to how](https://www.youtube.com/watch?v=hhUZ40ZWqkE) and transfer some PDEX tokens to the main account

Reviewer 3: Here is an instruction to how >> Here are the instructions

Reviewer 2: ```suggestion
[Create two accounts](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fblockchain.polkadex.trade#/accounts) [Here is one instruction video how to create accounts] (https://www.youtube.com/watch?v=hhUZ40ZWqkE) and transfer some PDEX tokens to the main account.
```

![Bonding Preferences](./screenshots/bonding_preferences.png)

- **Stash account** - Select your Stash account. In this example, we will bond 100 PDEX, where the minimum bonding amount is 100. Make sure that your Stash account contains at least this much. You can, of course, stake more than this.
- **Controller account** - Select the Controller account created earlier. This account will also need a small amount of PDEX in order to start and stop validating.

> - **Value bonded** - How much PDEX from the Stash account you want to bond/stake. Note that you do not need to bond all of the PDEX in that account. Also note that you can always bond more PDEX later. However, withdrawing any bonded amount requires the duration of the unbonding period. On Polkadex, the planned unbonding period is 28 days.

Reviewer 3: Question mark at the end of the first sentence

Reviewer 2: ```suggestion
- **Value bonded** - How much PDEX from the Stash account you want to bond/stake? You do not need to bond all of the PDEX in that account. Also note that you can always bond more PDEX later. However, withdrawing any bonded amount requires the duration of the unbonding period. On Polkadex, the planned unbonding period is 28 days.
```
